lib/ancestry/optimizers/pg.rb

```diff
@@ -0,0 +1,27 @@
+module Ancestry
+  module Optimizers
+    module Pg
+      # Update descendants with new ancestry (before save)
+      def update_descendants_with_new_ancestry
+        # If enabled and node is existing and ancestry was updated and the new ancestry is sane ...
+        if !ancestry_callbacks_disabled? && !new_record? && ancestry_changed? && sane_ancestry?
+          ancestry_column = ancestry_base_class.ancestry_column
+          old_ancestry = path_ids_in_database.join(Ancestry::MaterializedPath::ANCESTRY_DELIMITER)
+          new_ancestry = path_ids.join(Ancestry::MaterializedPath::ANCESTRY_DELIMITER)
+          update_clause = [
+            "#{ancestry_column} = regexp_replace(#{ancestry_column}, '^#{old_ancestry}', '#{new_ancestry}')"
+          ]
+
+          if ancestry_base_class.respond_to?(:depth_cache_column) && respond_to?(ancestry_base_class.depth_cache_column)
+            depth_cache_column = ancestry_base_class.depth_cache_column.to_s
+            update_clause << "#{depth_cache_column} = length(regexp_replace(regexp_replace(ancestry, '^#{old_ancestry}', '#{new_ancestry}'), '\\d', '', 'g')) + 1"
+          end
+
+          unscoped_descendants.update_all update_clause.join(', ')
```

Comment:
1. if we're doing this in sql, then locally cached models will be unchanged (and wrong).
2. update hooks are not called.

You will notice that `depth_cache_column` is reimplemented here. (we need a test that this is updated for children)

Another example: A model that stores the names of all the ancestors will have something like a `before_save` hook that changes the ancestor_names field. The update statement will need to be updated for these records.

All in all, this has similar tradeoffs as `destroy` vs `delete`.

Wonder if there is a mechanism that models can use to enhance that update clause.
Thinking that depth_cache_column needs to also use this mechanism rather than modifying this method.

When I finally introduce  `ancestor_ids` array column, I'll run into a similar optimization situation.